docs/src/proposals/scrambled-tx.md

```diff
@@ -0,0 +1,335 @@
+---
+title: Scrambled Transaction to combat against MEV
+---
+
+### Program Statement
+
+The existence of MEV isn't ideal and costly.
+
+MEV motivates transaction reordering, censoring and copying. Also, it causes
+cluster destabilization because of incentivised forking.
+
+The root problem is that there is no rule in the pursuit of MEV frontier, except
+economic rationality. While this anarchism is welcomed, stemmed from the
+blockchain philosophy, it's also desirable for such an activity to economically
+impractical in the first place.
+
+Any temporal market imbalance should be corrected strictly on the grounds of a
+fair game, basically the golden first-come-first-served rule (with optional
+bidding for priority).
+
+Note that this proposal doesn't fully mitigate one of MEV types, unlike all
+other fully-neutralized types: back-running. Back-running can be exercised
+while the golden rule is strictly and fairly enforced.
+
+### Summary of Proposed Solution
+
+This proposal tries to combat against MEV, at the cost of (slight) global
+induced latency of cluster as a whole.
+
+At highest level, a full mitigation of MEV can be attained _if leaders (and
+searchers) can't know the contents of other's transactions until finishing to
+packing and ordering them into a new block_. In addition to this, the
+transaction details must also be deterministically revealed after the ordering
+of transactions is committed.
+
+So, naive 2-step transaction submission (encrypted tx, then its key) won't work,
+let alone increased latency.
+
+Instead, this proposal uses a special transaction construction _only to delay
+the revelation of its content for the short amount of time tied to confirmation
+by the cluster_. Specifically, transactions are encrypted by a private key
+derived from the recursive sha256, and its seed and cipher text are submitted to
+the cluster at the same time.
+
+This construction is called *scrambling* across this proposal.
+
+Conversely, the forced computation to reveal the transaction details by solving
+the recursive sha256 is called *descrambling*.
+
+### Summary of consensus changes
+
+In addition to the obvious change of submission of scrambled transactions by
+clients, this proposal introduces a new pipeline stage (called the descrambling
+stage) and splits the voting into 2 phases (pre-vote and vote).
+
+Firstly, the leader feeds incoming transactions down the turbine immediately
+with minimum sanitization (This assumes leaders are now so-called bank-less and
+scheduling is delayed and on-chain via mempool).
+
+The validators pre-vote on the completed block, once they recover all data.
+Thus, the opt-conf commitment of data availability of the block can be rather
+established rather quickly, without replay timing variation by all other
+validators (This average duration will be the basis of difficulty adjustment of
+scrambling).
```

Comment:
> without replay timing variation by all other
validators (This average duration will be the basis of difficulty adjustment of
scrambling).

Are you saying that the difficulty of decrypting will depend on how fast validators were able to actually decrypt and apply changes the encrypted block contains during descrambling stage ? Do validators have any incentives to "reveal" the real decrypting power (with ASICs they have) to the network as opposed to just publishing the results at the latest possible moment ?
- if validators don't have these incentives it seems rational validators will just hide their real decrypting power and use it to try to extract MEV
- on the other hand, if validators do have those incentives for these PoW miners ... those incentives must be high enough to create and maintain dedicated ASICs specifically for that purpose (if incentives are low, a single malicious actor can attack network to stall it as he pleases)

Yet, it seems dynamic difficulty adjustment is necessary because if you set it to a constant value you can't really protect network from sudden increase of PoW decrypting power.

Reply:
well, scrambled tx is resistant to asics in general. that's because scrambling is single-threaed by repeated self-encryption.

so, the world's fastest descrabmler would be AES-NI. That means any custom hardware have to beat both amd/intel for single-core clock speed. I'd say it's impractical.

> Yet, it seems dynamic difficulty adjustment is necessary because if you set it to a constant value you can't really protect network from sudden increase of PoW decrypting power.

hehe, the proposal is quite outdated due to my laziness. I'm currently thinking about hard-coded constant drawn from some realistic jitter of block inclusion: https://github.com/solana-labs/solana/pull/23837#issuecomment-1732494143

Reply:
> Are you saying that the difficulty of decrypting will be adjusted dynamically depending on how fast validators were able to actually decrypt and apply changes the encrypted block contains during descrambling stage in the past

I think I got this ^ wrong, looks like you meant "the difficulty of decrypting" is determined based on the time it takes pre-vote stage to complete - such that the time it takes to decrypt **must be higher** than pre-vote stage takes, so that no user transactions can be decrypted during pre-vote stage (is there an upper bound on how long **pre-vote stage** can take ?), is that right @ryoqun (perhaps worth clarifying/expanding on this in prop text) ?

Another consideration is that you can descramble transactions is parallel (using a GPU or ASICs to speed things up). Looks like we get the best of both worlds!